### DISCUSSION TITLE
Unexpected Behavior of Tabulated Dihedral Angles

### DISCUSSION THREAD

**Discussion by Marsupail** [espressomd/espresso#4960] [Q&A] [answered]
Hello,
I am currently trying to implement a coarse grained single strand DNA model into espresso for a student project. The model uses tabulated potentials for the bonds, bond angles, dihedral angles and non bonded interactions. While trying to reproduce the probability distributions of the bonds and angles I was not able to reproduce them for the dihedral angles.

Then I tried to better understand the tabulated dihedral feature of espresso. For that I used the formula that is used for the Dihedral potential in espresso:
$$V(\phi) = 20(1-\cos(\phi-\pi)$$
This potential will result in a distribution with two almost equally large maxima: one at $0$ and one at $\pi$ instead of a distribution with only one maximum. 
**Potential:**
<img src="https://github.com/user-attachments/assets/4ad591b7-7394-4948-890c-0706e042ad96" width="400">

**Resulting Distribution:**
<img src="https://github.com/user-attachments/assets/d25aff2f-b134-4083-9322-1f76d9fa5c77" width="400">

If I use the following bond instead: `espressomd.interactions.Dihedral(bend=20, mult=1, phase=np.pi)` which should use the same formula as above I get the following distribution instead with one pronounced maximum at the minimum of the potential:
<img src="https://github.com/user-attachments/assets/3b30351b-09c2-4dfe-a81e-0750b7f99545" width="400">

My suspicion is, that espresso assumes the given potential is defined on $[0,\pi)$ and then extends it periodically to get the full interval but this would be contrary to the documentation.  

Is this the expected behavior and did I miss something in the documentation? And if so how would I define a asymmetric tabulated dihedral potential?

Another question I had is, is the force for a tabulated dihedral potential expected to be the negative derivative of the energy or the positive one (since for the tabulated angle potential it is expected to be the positive derivation)?

Best regards and thanks for the help,
Emelie

**Comment by jngrad** [comment]
Hi,

First of all, congratulations for double-checking the tabulated potential against a reference potential. Poorly documented features like `TabulatedDihedral` can be quite dangerous, since they do not necessarily work in the way most users expect them to work. Verifying individual components of a Hamiltonian when building a simulation script is good practice.

The dihedral potential is defined on the inclusive range $\left[0, 2\pi\right]$. From what I can tell, by looking at the source code, the tabulated force should evaluate the expression $A(\phi) = -nK\sin(n \phi - \phi_0) / \sin(\phi)$ outside the singularities. I would encourage you to double-check my claim, as I'm not 100% sure and am not in contact with the original author of this feature.

Using [testsuite/python/interactions_dihedral.py@`8ee6f00`](https://github.com/espressomd/espresso/blob/8ee6f00/testsuite/python/interactions_dihedral.py) to derive the expressions of the tabulated potential and force, I get the following results for `Dihedral` ("classic") and `TabulatedDihedral` ("tabulated"):

![Potential energy surface: the tabulated dihedral potential overlaps perfectly with the classic dihedral potential.](https://github.com/user-attachments/assets/127ea50d-bf1f-4f65-803f-0c71a337bc0f)
![Force deviation: the tabulated dihedral force error is in the range 1e-15 compared to the classic dihedral force.](https://github.com/user-attachments/assets/e168d523-3809-4dc6-b7cc-8174d493ce33)

The forces on particles 1 and 3 are obtained by symmetry from the forces on particles 0 and 2, and are thus not represented. The deviation is less than $10^{-14}$ in reduced units for a tabulated dihedral potential with a resolution of 40 points. The deviation is undefined at the two singularities ($\pi$ and $2\pi$), although due to interpolation they may become finite.

Your first histogram is quite puzzling to me. Maybe your definition of the tabulated dihedral force doesn't match ESPResSo's internal definition. Unfortunately, there aren't many users for the `TabulatedDihedral` potential, and this is why it was never properly documented.

> Another question I had is, is the force for a tabulated dihedral potential expected to be the negative derivative of the energy or the positive one (since for the tabulated angle potential it is expected to be the positive derivation)?

For dihedrals the negative derivative is used. Also, when the multiplicity of the energy is $n$, the multiplicity of the force is $n-1$. It is not immediately obvious when looking at the equations of the dihedral forces, since they involve a lot of cross products, but you can convince yourself with a graphical solution. For $n = 1$ the force is $F(\theta) = K$ in reduced units (flat profile). For $n = 2$, the force is $F(\theta) = 4K\cdot\cos(\theta)$, as shown below:

![Superposition of the force and energy profile of a dihedral.](https://github.com/user-attachments/assets/3cbe2994-cd43-4220-859b-5f5720a4320a)

Please note it is not possible to make further deductions from this plot. In particular, the fact that the tabulated force $A(\phi)$ is zero at the minima of the tabulated energy is a coincidence and doesn't hold for other values of $n$. There might be an elegant proof for the $n - 1$ periodicity of $A(\phi)$, although I'm not aware of one. If this detail is of interest to you, you may want to look at mathematical proofs for quotients of sines. For products of sines, there are strategies available, but quotients are more challenging. Maybe identities of the Fourier transform would also help.

The script to generate these plots is reproduced below. I hope this was helpful.

<details><summary>potential energy surface script (click to unroll)</summary>

```python
import espressomd
import espressomd.interactions
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker

plt.rcParams.update({"font.size": 12})
radians_formatter = mticker.FuncFormatter(lambda x, _: rf"${x / np.pi:.0f}\pi$" if x != 0. else "0")
radians_locator = mticker.MultipleLocator(base=np.pi)

system = espressomd.System(box_l=[10.0, 10.0, 10.0])
system.cell_system.skin = 0.4
system.time_step = 0.1

def potential_energy_surface(bond, resolution=60):
    p0 = system.part.add(pos=[5., 6., 5.])
    p1 = system.part.add(pos=[5., 5., 5.])
    p2 = system.part.add(pos=[6., 5., 5.])
    p3 = system.part.add(pos=[5., 6., 5.])
    p1.add_bond((bond, p0, p2, p3))
    v = np.array([0., 1., 0.])
    k = np.array([1., 0., 0.])
    xdata = np.linspace(0., 2. * np.pi, num=resolution, endpoint=True)
    ydata = []
    zdata = []
    for phi in xdata:
        p3.pos = p2.pos + v * np.cos(phi) + np.cross(k, v) * np.sin(phi) + k * np.dot(k, v) * (1. - np.cos(phi))
        system.integrator.run(recalc_forces=True, steps=0)
        ydata.append(system.analysis.energy()["bonded"])
        zdata.append([p0.f, p1.f, p2.f ,p3.f])
    system.part.clear()
    ydata = np.array(ydata)
    zdata = np.array(zdata)
    return (xdata, ydata, zdata)

def make_tabulated(dihedral, N):
    assert N % 2 == 0
    phi = np.linspace(0., 2. * np.pi, num=N + 1, endpoint=True)
    tab_energy = dihedral.bend * (1. - np.cos(dihedral.mult * phi - dihedral.phase))
    div = np.sin(phi)
    div[0] = div[N // 2] = div[N] = 1.
    tab_force = -dihedral.bend * dihedral.mult * np.sin(dihedral.mult * phi - dihedral.phase) / div
    tab_force[0] = tab_force[N // 2] = tab_force[N] = 0.
    dihedral_tabulated = espressomd.interactions.TabulatedDihedral(
        energy=tab_energy, force=tab_force)
    return dihedral_tabulated

dihedral = espressomd.interactions.Dihedral(bend=20., mult=1, phase=np.pi)
dihedral_tab = make_tabulated(dihedral, 40)
system.bonded_inter.add(dihedral)
system.bonded_inter.add(dihedral_tab)

xdata, ydata, zdata = potential_energy_surface(dihedral)
xdata, ydata_tab, zdata_tab = potential_energy_surface(dihedral_tab)

# plot potential energy surface
fig, ax = plt.subplots(figsize=(5, 3))
ax.plot(xdata, ydata, "-", label="classic")
ax.plot(xdata, ydata_tab, "o", mfc="none", label="tabulated")
ax.xaxis.set_major_formatter(radians_formatter)
ax.xaxis.set_major_locator(radians_locator)
ax.xaxis.set_label_text("Dihedral angle")
ax.yaxis.set_label_text("Energy (reduced units)")
ax.set_title("Potential energy surface")
ax.legend(loc="upper center")
fig.tight_layout()
plt.show()

# plot force deviations
fig, ax = plt.subplots(figsize=(5, 3))
for i in [0, 2]:
    diff = np.linalg.norm(zdata[:,i] - zdata_tab[:,i], axis=1)
    N = len(diff)
    # erase singularities at 0, pi, 2*pi
    for j in [0, 1, N // 2 - 1, N // 2, N - 2, N - 1]:
        diff[j] = np.nan
    ax.plot(xdata, diff, "-", label=rf"$\Delta F_{i}$")

ax.xaxis.set_major_formatter(radians_formatter)
ax.xaxis.set_major_locator(radians_locator)
ax.xaxis.set_label_text("Dihedral angle")
ax.yaxis.set_label_text("Force deviation (reduced units)")
ax.set_title("Numerical error in the forces")
ax.legend(loc="upper center")
fig.tight_layout()
plt.show()

# plot multiplicity of the derivative
dihedral = espressomd.interactions.Dihedral(bend=20., mult=2, phase=np.pi)
dihedral_tab = make_tabulated(dihedral, 180)
tab_energy = np.copy(dihedral_tab.energy)
tab_force = np.copy(dihedral_tab.force)
N = len(tab_force)
tab_phi = np.linspace(0., 2. * np.pi, num=N, endpoint=True)
for j in [0, 1, N // 2 - 1, N // 2, N - 2, N - 1]:
    tab_force[j] = np.nan

fig, ax = plt.subplots(figsize=(5, 3))
ax.plot(tab_phi, tab_force, label="force")
ax.plot(tab_phi, tab_energy, label="energy")
ax.xaxis.set_major_formatter(radians_formatter)
ax.xaxis.set_major_locator(radians_locator)
ax.yaxis.set_major_locator(mticker.MultipleLocator(base=2. * dihedral.bend))
ax.xaxis.set_label_text("Dihedral angle")
ax.yaxis.set_label_text("Functional (reduced units)")
ax.legend(loc="upper center")
ax.spines["bottom"].set_position(("data", 0))
ax.spines["top"].set_visible(False)
ax.spines["right"].set_visible(False)
fig.tight_layout()
plt.show()
```

</details>

**Reply by Marsupail** [reply]
Yes thank you very much for the detailed answer and the script 😄. My simulation behaves differently since my force is indeed different. I used:
$$A(\phi)=-nK\sin(n\phi-\phi_0)$$ 
without the $\sin(\phi)$ in the divisor. But for me it is not clear where this divisor comes from. I thought for the force I just need to calculate $A(\phi)= -\partial_{\phi} V(\phi)$ and this should result in the equation above.
So if I wan't to use a arbitrary tabulated potential I would have to set the force to:
$$A(\phi)= -\partial_{\phi} V(\phi)/\sin(\phi) $$
where I calculate the derivation numerically?

**Reply by jngrad** [reply, marked as answer]
> So if I wan't to use a arbitrary tabulated potential I would have to set the force to:
>
>  $A(\phi) = -\partial_{\phi} V(\phi) / \sin(\phi)$
>
> where I calculate the derivation numerically?

I would say yes. The source code is a bit cryptic, so I cannot answer you with full certainty.

The term $1/\sin(\phi)$ is due to how the derivative is obtained. The quantity $-\partial_{\phi} V(\phi)$ yields the magnitude of the force, but one still needs to compute the gradient of the dihedral angle with respect to the particle coordinates, in order to get a force vector.

There are several ways of expressing the derivative. Using the chain rule, and depending on whether the source code evaluates $\cos(\phi)$ resp. $\phi$ as an intermediate value (the latter is less common, since arccos is computationally expensive), one can use equation (29) resp. (30). In the special case where $\phi_0=0$, the $1/\sin(\phi)$ term in (30) can be simplified out to yield (31).

$$\nabla_i V(\phi) = \left(\dfrac{d V(\phi)}{d \phi}\right)\cdot\nabla_i\phi \tag{29}$$

$$
\begin{align}
\nabla_i V(\phi) &= \left(\dfrac{d V(\phi)}{d\phi}\right)\left(\dfrac{d \phi}{d\cos(\phi)}\right)\cdot\nabla_i\cos(\phi) \\
  &= \left(\dfrac{d V(\phi)}{d\phi}\right)\left(\dfrac{-1}{\sin(\phi)}\right)\cdot\nabla_i\cos(\phi) \tag{30}
\end{align}
$$

$$\nabla_i V(\phi) = \left(\dfrac{d V(\phi)}{d\cos(\phi)}\right)\cdot\nabla_i\cos(\phi) \tag{31}$$

Therefore, whether we end up dividing by $\sin(\phi)$ or not depends on the functional form of $V(\phi)$. ESPResSo implements the general case, which is why there are two singularities. There are other ways to express the derivative, for example the CHARMM force field introduces an extra intermediate variable $\sin(\phi)$ in the source code, thus avoiding singularities in the general case.

Equation numbers come from Swope et Ferguson, *Alternative expressions for energies and forces due to angle bending and torsional energy*, in: Journal of Computational Chemistry 13(5):585–594, 1992, doi:[10.1002/jcc.540130508](https://doi.org/10.1002/jcc.540130508). The frequent jumps from polar to Cartesian coordinates can be a bit confusing in the beginning. I found it helpful to look at the chain rules for bond angle potentials (section [Bond angle potentials](https://espressomd.github.io/dox/bondedIA.html#bondedIA_bond_angles) in the ESPResSo Doxygen documentation), before investigating dihedral potentials. For completeness, I should also mention Blondel et Karplus, *New formulation for derivatives of torsion angles and improper torsion angles in molecular mechanics: Elimination of singularities*, in: Journal of Computational Chemistry 17(9):1132-1141, 1996, doi:[10.1002/(SICI)1096-987X(19960715)17:9<1132::AID-JCC5>3.0.CO;2-T](https://doi.org/10.1002/(SICI)1096-987X(19960715)17:9<1132::AID-JCC5>3.0.CO;2-T), where the dihedral force is expressed without singularities for the general case.

**Reply by Marsupail** [reply]
Ah okay. Thank you very much for the detailed answers. With the added divisor my results are way closer to the target. And also thank you for the sources I will read into it to better understand whats done here.
Maybe it would be good to mention the divisor somewhere in the documentation? I think it is not obvious that it is needed.
